### ISSUE TITLE
When using in MasterDetail OnNavigatedTo is always raised in Master Detail

### ISSUE THREAD

**Issue by mackayn** [dansiegel/Prism.Plugin.Popups#16]
In MasterDetail apps when you Push a Popup page from the view model of the detail and go back OnNavigatedTo does not trigger in the viewmodel that called the code below (ViewA), instead OnNavigatedTo triggers in the view model of the MasterDetail page

```await _navigationService.PushPopupPageAsync(PageLibrary.PopupPageKeyFleet);```

The App structure is

MasterDetail\NavigationPage

- MasterDetailPage\NavigationPage\ViewA
- MasterDetailPage\NavigationPage\ViewB
- MasterDetailPage\NavigationPage\ViewC

etc

I got round it for a demo using the event aggregator but I don't think this is the intended behavior. I only noticed when we switched out app to use MasterDetail.

I tested pushing a regular modal page and OnNavigatedTo does trigger in the viewmodel of ViewA.

I tested with 1.1.0-pre1 & 1.0.1 stable and both version has the same issue.

**Comment by dansiegel** [comment]
Just to confirm the behavior you're looking for would be to call OnNavigatedTo on the Detail of the MasterDetailPage and if that page is a NavigationPage, then the CurrentPage of the NavigationPage. Is that correct?

**Comment by mackayn** [comment]
Hi Dan,

Yes, I believe so. It works fine in the Navigation page scenario, it's the in master detail scenario the issue occurs.

**Comment by dansiegel** [comment]
Pre2 is now available with this fix

**Comment by mackayn** [comment]
HI Dan.
Thanks for the fix, this version won't show a Popup, when you call PushPopupPageAsync it just hangs the app.

```
01-20 10:00:27.248 D/ViewRootImpl(16382): ViewPostImeInputStage processPointer 0
01-20 10:00:27.448 D/ViewRootImpl(16382): ViewPostImeInputStage processPointer 1
Resolved pending breakpoint at 'HomePageViewModel.cs:95,1' to void MarineOps.ViewModels.HomePageViewModel.<ShowAddActions>d__17.MoveNext () [0x00008].
01-20 10:00:27.508 D/Mono    (16382): Assembly Ref addref MarineOps[0xaa090ae0] -> Prism.Plugin.Popups.Unity[0xadded280]: 2
01-20 10:00:27.513 D/Mono    (16382): Assembly Loader probing location: '/storage/emulated/0/Android/data/com.vships.marineops/files/.__override__/System.Threading.Tasks.dll'.
01-20 10:00:27.513 D/Mono    (16382): AOT module '/storage/emulated/0/Android/data/com.vships.marineops/files/.__override__/System.Threading.Tasks.dll.so' not found: dlopen failed: library "/data/app/com.vships.marineops-1/lib/arm/libaot-System.Threading.Tasks.dll.so" not found
01-20 10:00:27.513 D/Mono    (16382): AOT module '/Users/builder/data/lanes/3511/501e63ce/source/monodroid/builds/install/mono-armv7/lib/mono/aot-cache/arm/System.Threading.Tasks.dll.so' not found: dlopen failed: library "/data/app/com.vships.marineops-1/lib/arm/libaot-System.Threading.Tasks.dll.so" not found
01-20 10:00:27.513 D/Mono    (16382): Assembly Ref addref Prism.Plugin.Popups.Unity[0xadded280] -> System.Threading.Tasks[0xaa0910e0]: 12
01-20 10:00:27.513 D/Mono    (16382): Assembly Ref addref Prism.Plugin.Popups.Unity[0xadded280] -> Prism.Forms[0xadded400]: 13
01-20 10:00:30.073 D/Mono    (16382): Assembly Ref addref Prism.Plugin.Popups.Unity[0xadded280] -> Xamarin.Forms.Core[0xaa0905a0]: 31
01-20 10:00:30.073 D/Mono    (16382): Assembly Ref addref Prism.Plugin.Popups.Unity[0xadded280] -> Rg.Plugins.Popup[0xadded5e0]: 4
01-20 10:00:30.078 D/Mono    (16382): Assembly Loader probing location: '/storage/emulated/0/Android/data/com.vships.marineops/files/.__override__/System.Collections.dll'.
01-20 10:00:30.083 D/Mono    (16382): AOT module '/storage/emulated/0/Android/data/com.vships.marineops/files/.__override__/System.Collections.dll.so' not found: dlopen failed: library "/data/app/com.vships.marineops-1/lib/arm/libaot-System.Collections.dll.so" not found
01-20 10:00:30.083 D/Mono    (16382): AOT module '/Users/builder/data/lanes/3511/501e63ce/source/monodroid/builds/install/mono-armv7/lib/mono/aot-cache/arm/System.Collections.dll.so' not found: dlopen failed: library "/data/app/com.vships.marineops-1/lib/arm/libaot-System.Collections.dll.so" not found
01-20 10:00:30.083 D/Mono    (16382): Assembly Ref addref Rg.Plugins.Popup[0xadded5e0] -> System.Collections[0xaa090900]: 16
01-20 10:00:30.083 D/Mono    (16382): Assembly Loader probing location: '/storage/emulated/0/Android/data/com.vships.marineops/files/.__override__/System.Threading.Tasks.dll'.
01-20 10:00:30.083 D/Mono    (16382): AOT module '/storage/emulated/0/Android/data/com.vships.marineops/files/.__override__/System.Threading.Tasks.dll.so' not found: dlopen failed: library "/data/app/com.vships.marineops-1/lib/arm/libaot-System.Threading.Tasks.dll.so" not found
01-20 10:00:30.083 D/Mono    (16382): AOT module '/Users/builder/data/lanes/3511/501e63ce/source/monodroid/builds/install/mono-armv7/lib/mono/aot-cache/arm/System.Threading.Tasks.dll.so' not found: dlopen failed: library "/data/app/com.vships.marineops-1/lib/arm/libaot-System.Threading.Tasks.dll.so" not found
01-20 10:00:30.083 D/Mono    (16382): Assembly Ref addref Rg.Plugins.Popup[0xadded5e0] -> System.Threading.Tasks[0xaa0910e0]: 13
01-20 10:00:30.088 D/Mono    (16382): Assembly Ref addref Prism.Plugin.Popups.Unity[0xadded280] -> Microsoft.Practices.Unity[0xaddecb00]: 12
01-20 10:00:30.088 D/Mono    (16382): Assembly Ref addref Prism.Plugin.Popups.Unity[0xadded280] -> Prism.Unity.Forms[0xadded580]: 8
01-20 10:00:30.128 D/Mono    (16382): Assembly Loader probing location: '/storage/emulated/0/Android/data/com.vships.marineops/files/.__override__/System.Runtime.Extensions.dll'.
01-20 10:00:30.128 D/Mono    (16382): AOT module '/storage/emulated/0/Android/data/com.vships.marineops/files/.__override__/System.Runtime.Extensions.dll.so' not found: dlopen failed: library "/data/app/com.vships.marineops-1/lib/arm/libaot-System.Runtime.Extensions.dll.so" not found
01-20 10:00:30.128 D/Mono    (16382): AOT module '/Users/builder/data/lanes/3511/501e63ce/source/monodroid/builds/install/mono-armv7/lib/mono/aot-cache/arm/System.Runtime.Extensions.dll.so' not found: dlopen failed: library "/data/app/com.vships.marineops-1/lib/arm/libaot-System.Runtime.Extensions.dll.so" not found
01-20 10:00:30.128 D/Mono    (16382): Assembly Ref addref Rg.Plugins.Popup[0xadded5e0] -> System.Runtime.Extensions[0xaa090f60]: 7
01-20 10:00:30.133 D/Mono    (16382): Assembly Loader probing location: '/storage/emulated/0/Android/data/com.vships.marineops/files/.__override__/System.Reflection.dll'.
01-20 10:00:30.133 D/Mono    (16382): AOT module '/storage/emulated/0/Android/data/com.vships.marineops/files/.__override__/System.Reflection.dll.so' not found: dlopen failed: library "/data/app/com.vships.marineops-1/lib/arm/libaot-System.Reflection.dll.so" not found
01-20 10:00:30.133 D/Mono    (16382): AOT module '/Users/builder/data/lanes/3511/501e63ce/source/monodroid/builds/install/mono-armv7/lib/mono/aot-cache/arm/System.Reflection.dll.so' not found: dlopen failed: library "/data/app/com.vships.marineops-1/lib/arm/libaot-System.Reflection.dll.so" not found
01-20 10:00:30.133 D/Mono    (16382): Assembly Ref addref Rg.Plugins.Popup[0xadded5e0] -> System.Reflection[0xaa090cc0]: 15
01-20 10:00:30.133 D/Mono    (16382): Assembly Ref addref Rg.Plugins.Popup[0xadded5e0] -> System.Reflection.Extensions[0xaa091020]: 11
01-20 10:00:30.138 D/Mono    (16382): Assembly Loader probing location: '/storage/emulated/0/Android/data/com.vships.marineops/files/.__override__/System.Linq.dll'.
01-20 10:00:30.138 D/Mono    (16382): AOT module '/storage/emulated/0/Android/data/com.vships.marineops/files/.__override__/System.Linq.dll.so' not found: dlopen failed: library "/data/app/com.vships.marineops-1/lib/arm/libaot-System.Linq.dll.so" not found
01-20 10:00:30.138 D/Mono    (16382): AOT module '/Users/builder/data/lanes/3511/501e63ce/source/monodroid/builds/install/mono-armv7/lib/mono/aot-cache/arm/System.Linq.dll.so' not found: dlopen failed: library "/data/app/com.vships.marineops-1/lib/arm/libaot-System.Linq.dll.so" not found
01-20 10:00:30.138 D/Mono    (16382): Assembly Ref addref Rg.Plugins.Popup[0xadded5e0] -> System.Linq[0xaa090fc0]: 11
01-20 10:00:30.173 D/Mono    (16382): Assembly Loader probing location: '/storage/emulated/0/Android/data/com.vships.marineops/files/.__override__/System.Linq.dll'.
01-20 10:00:30.173 D/Mono    (16382): AOT module '/storage/emulated/0/Android/data/com.vships.marineops/files/.__override__/System.Linq.dll.so' not found: dlopen failed: library "/data/app/com.vships.marineops-1/lib/arm/libaot-System.Linq.dll.so" not found
01-20 10:00:30.173 D/Mono    (16382): AOT module '/Users/builder/data/lanes/3511/501e63ce/source/monodroid/builds/install/mono-armv7/lib/mono/aot-cache/arm/System.Linq.dll.so' not found: dlopen failed: library "/data/app/com.vships.marineops-1/lib/arm/libaot-System.Linq.dll.so" not found
01-20 10:00:30.173 D/Mono    (16382): Assembly Ref addref Prism.Plugin.Popups.Unity[0xadded280] -> System.Linq[0xaa090fc0]: 12
01-20 10:00:38.448 D/Mono    (16382): [0x8b22d930] worker finishing
Thread finished: <Thread Pool> #5
The thread 'Unknown' (0x5) has exited with code 0 (0x0).
01-20 10:00:44.273 D/Mono    (16382): [0x89b3d930] worker finishing
Thread finished: <Thread Pool> #11
The thread 'Unknown' (0xb) has exited with code 0 (0x0).

```
No exception is raised and the task never returns.

I have uploaded a repo project for you

https://drive.google.com/drive/folders/0B90tKD37uOtwZ3pjSDlTMzR3Z2M?usp=sharing

I think the issue is due to the detail been wrapped in a NavigationPage possibly. This is a fairly common scenario in Prism Master Detail.

Sorry, I didn't make that clear in my earlier posts.

**Comment by nglekhoi** [comment]
Modify the FilterPage function inside PopupExtensions.cs like this:
```csharp
         private static Page FilterPage(Page page)
        {
            if (page == null) return page;
            var startPage = page;

            while (page is MasterDetailPage || page is NavigationPage || page.GetType() == typeof(MultiPage<>))
            {
                if (page.GetType().GetTypeInfo().IsSubclassOf(typeof(MultiPage<>)))
                {
                    page = (page as MultiPage<Page>).CurrentPage;
                }
                else if (page.GetType().GetTypeInfo().IsSubclassOf(typeof(MasterDetailPage)))
                {
                    page = (page as MasterDetailPage).Detail;
                }
                else if (page.GetType().GetTypeInfo().IsSubclassOf(typeof(NavigationPage)))
                {
                    page = (page as NavigationPage).CurrentPage;
                }
            }

            return page ?? startPage;
        }
```
fixed the Master Detail problem for me. 

The idea is using
`
page.GetType().GetTypeInfo().IsSubclassOf(typeof(MasterDetailPage)))
`
instead of
`
page.GetType() == typeof( MasterDetailPage )
`

Because if you compare the Type of derived class of MasterDetailPage to MasterDetailPage, the result always false. So you will get the infinite loop, which describe why the app hang.

Please fix and upload new package to nuget, thanks.

**Comment by mackayn** [comment]
I did notice an unwanted behavior though in the following scenario in the latest build.

If I'm call the Popup from MasterDetails\NavigationPage\Homepage,  Destroy in the homepage Viewmodel is called.

Homepage is rendered and the current detail page so really it shouldn't be destroyed as it's in the navigation stack.

So the current detail page will do it's cleanup although it's the current detail page when the popup page closes.

**Comment by mackayn** [comment]
Hi,

I just tested with the sample app in github and it's the same issue with just a single contentpage, Destroy is called when you show the popup.

```
using System.Diagnostics;
using Prism.Navigation;
using Xamarin.Forms;

namespace PopupPluginSample.Views
{
    public partial class MainPage : ContentPage, IDestructible
    {
        public MainPage()
        {
            InitializeComponent();
        }

        public void Destroy()
        {
            Debug.WriteLine("MainPage destroyed");
        }
    }
}
```

I'll step through and see what's going on.

**Comment by mackayn** [comment]
Hi,

The issues seems to be in HandleINavigatedAware, if you handling navigating from HandleIDestructiblePage is called even if it's not a PopupPage, this is the job of Prism's navigation service though.

The change I think will resolve this issue would be:

Original code:
```
                pageAware?.OnNavigatedFrom( parameters );
                contextAware?.OnNavigatedFrom( parameters );
                HandleIDestructiblePage( page );
```

Change

```
                pageAware?.OnNavigatedFrom( parameters );
                contextAware?.OnNavigatedFrom( parameters );
                if (page is PopupPage) HandleIDestructiblePage( page );
```

That way only the popup page will be cleaned up  but OnNavigatedFrom will be triggered correctly.

**Comment by dansiegel** [comment]
Should be fixed by PR #25
